### DISCUSSION TITLE
WiFi.h: No such file......

### DISCUSSION THREAD

**Discussion by mgaman** [pschatzmann/arduino-audio-tools#1576] [Q&A] [answered]
Has that happened to you?
When compiling the example  examples-communication/a2dp/player-sdfat-a2dp on Arduino IDE everything went fine and it just worked, without a single change needed.
However when I moved over to VSCode/PlatformIO things got a bit weird. Libraries were loaded by adding them to the lib_deps section of platformio.ini. The contents of player-sdfat-a2dp.ino source file were copied to main.cpp and the app compiled. Thats when I got the Missing WiFi.h file message.
The problem only occurred on a2dp examples. player-sdfat-audiokit compiled and ran just fine. Knowing that on ESP32 you cannot physically execute both Bluetooth and WiFi together, as there is only one physical RF device I guessed that this was the root of the problem.
In the AudioConfig.h file there is a whole list of options. I found the following 4 lines that are WiFi related
#define USE_WIFI
#define USE_WIFI_CLIENT_SECURE
#define USE_URL_ARDUINO
#define USE_AUDIO_SERVER
For the player-sdfat-a2dp example I commented out the above 4 lines. Now everthing works. It compiles without any errors and executes correctly.
I am told that PlatformIO compiles every source file, including libraries, and later has a filtering mechanism to select just the files it needs for linkage. If Arduino IDE only compiles the source files needed it would explain the different behaviour of the two platforms.
Whatever the reason, I'm glad to share this with your users. I'm sure that I'm not the first to encounter the problem.

**Comment by pschatzmann** [comment, marked as answer]
Wouldn't you rather file an issue or a question with PlatformIO ?

I googled your issue and it seems that adding lib_ldf_mode = chain+ to the platformio.ini will make the issue go away

The following is working for me for your example 
```
[env:esp32dev]
platform = espressif32
board = esp32dev
framework = arduino
lib_deps = https://github.com/pschatzmann/arduino-audio-tools.git
           https://github.com/pschatzmann/arduino-libhelix
           https://github.com/pschatzmann/ESP32-A2DP
           https://github.com/greiman/SdFat

build_flags = -DCORE_DEBUG_LEVEL=5 -DFILE_COPY_CONSTRUCTOR_SELECT=2 -Wno-unused-variable -Wno-unused-but-set-variable -Wno-unused-function -Wno-format-extra-args 
monitor_speed = 115200
monitor_filters = esp32_exception_decoder
lib_ldf_mode = chain+
board_build.partitions = huge_app.csv

```

ps. The A2DPStream does not need any output functionality of the A2DP library, so you could switch them off with the build_flags 
 -DA2DP_I2S_AUDIOTOOLS=0 -DA2DP_LEGACY_I2S_SUPPORT=0

**Reply by mgaman** [reply]
Thanks for the prompt reply.
I adopted the "chain+" line and it worked fine.
The build_flag CORE_DEBUG_LEVEL=5 causes a compile error
<b>pio/libdeps/esp32dev/ESP32-A2DP/src/BluetoothA2DPSource.cpp:969:69: error: 'struct esp_avrc_ct_cb_param_t::avrc_ct_psth_rsp_param' has no member named 'rsp_code'; did you mean 'key_code'?</b>
If I understand the documentation correctly levels 0-2 are none/error/warning  while levels 3-5 are info/debug/verbose. 
Anyway, I prefer the chain+ solution to mine which changes the original source code.

**Comment by mgaman** [comment]
This is not really an issue with audio-tools or platformio, it just
illustrates my lack of detailed knowledge of the finer points of platformio.
However I cannot be the first nor certainly the last to trip up on this
problem.
Might I suggest that
https://github.com/pschatzmann/arduino-audio-tools/wiki/Working-with-PlatformIO
be updated?

**Comment by pschatzmann** [comment]
I actually see this issue the first time, so I tend to think that it might have been introduced quite lately by a change in PlatformIO.
CORE_DEBUG_LEVEL=5 works w/o any problems for me.

Here is the version information for you to double check

```
CONFIGURATION: https://docs.platformio.org/page/boards/espressif32/esp32dev.html
PLATFORM: Espressif 32 (6.6.0+sha.c9c9bbe) > Espressif ESP32 Dev Module
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
DEBUG: Current (cmsis-dap) External (cmsis-dap, esp-bridge, esp-prog, iot-bus-jtag, jlink, minimodule, olimex-arm-usb-ocd, olimex-arm-usb-ocd-h, olimex-arm-usb-tiny-h, olimex-jtag-tiny, tumpa)
PACKAGES: 
 - framework-arduinoespressif32 @ 3.20014.231204 (2.0.14) 
 - tool-esptoolpy @ 1.40501.0 (4.5.1) 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain+, Compatibility ~ soft
Found 37 compatible libraries
Scanning dependencies...
Dependency Graph
|-- audio-tools @ 0.9.8+sha.7656898
|-- libhelix @ 0.1.0+sha.9574742
|-- ESP32-A2DP @ 1.8.1+sha.c866889
|-- SdFat @ 2.2.3+sha.aadedbf
|-- WiFi @ 2.0.0
|-- WiFiClientSecure @ 2.0.0
|-- SPI @ 2.0.0
Building in release mode
Compiling .pio/build/esp32dev/src/main.cpp.o
Building .pio/build/esp32dev/bootloader.bin
Generating partitions .pio/build/esp32dev/partitions.bin
esptool.py v4.5.1
Creating esp32 image...
Merged 1 ELF section
Successfully created esp32 image.
```

**Comment by mgaman** [comment]
Actually my environment seems to be (marginally) older than yours
PACKAGES:
 - framework-arduinoespressif32 @ 3.20007.0 (2.0.7)
 - tool-esptoolpy @ 1.40500.0 (4.5.0)

Everything else is the same.
I then compared my Arduino IDE and PlatformIO installations in ~/.arduino15
and ~/.platformio and noticed small differences, IDE being newer, like your
installation.
To force an update I executed pio pkg update so now I have
framework-arduinoespressif32 2.0.16 and esptool 4.5.1.
Hey presto! All compiles even with CORE_DEBUG_LEVEL=5

Well, thats a few hours of my life I won't get back. Never mind, all is
well now.

**Comment by vorosj** [comment]
In platformio.ini:
platform = espressif32@6.6.0 
or newer forces PIO to use  a known version.
